src/main/java/seedu/duke/command/CommandParser.java

```diff
@@ -0,0 +1,141 @@
+ * 3. PAYLOADS - value of the action/parameters <br>
+ * In short, user input is a list of commands, each command containing arguments and payloads. <br>
+ * <br>
+ * For example, a given user input: <code>"deadline work on CS2113 --by Sunday"</code>
+ * <li>Has commands ["deadline work on CS2113", "by Sunday"]</li>
+ * <li>Has arguments ["deadline", "by"]</li>
+ * <li>Has payloads ["work on CS2113", ["Sunday"]</li>
+ * <br>
+ * Further, we define the FIRST command to be the MAIN command of any given user input. <br>
+ * So, <code>"deadline work on CS2113 --by Sunday"</code> has <code>"deadline work on CS2113"</code>
+ * as the main command
+ */
+public class CommandParser {
+
+    private static final String ARGUMENT_DELIMITER = " --";
+    private static final String UNPADDED_DELIMITER = "--";
+    private static final String PAYLOAD_DELIMITER = " ";
+
+    // Message string constants for errors and ui
+    private static final String ERROR_EMPTY_COMMAND = "Command length is empty.";
+    private static final String ERROR_EMPTY_ARGUMENT = "Command is missing an argument!";
+
+    /**
+     * Constructs an instance of CommandParser. <br>
+     * <p>
+     * CommandParser should be used to break down raw user input into
+     * logical <code>[Argument,Payload]</code> pairs
+     */
+    public CommandParser() {
+
+    }
+
+    private String[] splitIntoCommands(String fullCommandString) throws BadCommandException {
+        // Perform a string length sanity check
+        fullCommandString = fullCommandString.strip();
+        if (fullCommandString.length() == 0) {
+            throw new BadCommandException(ERROR_EMPTY_COMMAND);
+        }
+
+        String[] commands = fullCommandString.split(ARGUMENT_DELIMITER, -1);
+        // Adversarial user input check
+        // There are 2 possible adversarial inputs that should be checked for
+        // 1. Whitespace/Empty Arguments: `cmd payload -- payload1 -- `
+        //    Split renders it as ["cmd payload", " payload1", ""]
+        //    " payload1" will cause issues with rendering
+        //    So, check for empty commands and whitespace prefix
+        // 2. Missing main argument: `--argument payload`
+        //    Split renders this as ["--argument payload"]
+        //    So, check for "--" prefix
+
+        String[] cleanCommands = new String[commands.length];
+        for (int i = 0; i < commands.length; ++i) {
+            String currentCommand = commands[i];
+            // Case 1 check
+            if (currentCommand.startsWith(" ") || currentCommand.length() == 0) {
+                throw new BadCommandException(ERROR_EMPTY_ARGUMENT);
+            }
+            currentCommand = currentCommand.strip();
+            // Case 2 check
+            if (currentCommand.startsWith(UNPADDED_DELIMITER)) {
+                throw new BadCommandException(ERROR_EMPTY_COMMAND);
+            }
+            cleanCommands[i] = currentCommand;
+        }
+        return cleanCommands;
+    }
```

Comment:
For case 1, what if the argument is totally missing? It fulfills none of the cases and will be mis-treated as a command instead in the subsequent processes?

Reply:
Can you give an example?   

Your phrasing "missing argument" is vague as it could mean that the delimiter is present but the name of the argument is not.
Or are you asking if there is no argument at all?  e.g. `mainCommand payload`

If it helps, I  accounted for various edge cases in the [unit test](src/test/java/seedu/duke/command/CommandParserTest.java) that encompass "missing argument"  with inputs such as: 
- `mainCommand payload --`
- `mainCommand payload -- `
- `mainCommand payload -- payload1`
^ All of the above will throw an error as the argument is 'missing'.

Reply:
let say:
```
mainCommand payload --payload1
```
there is no space between '--' and 'payload1'
so what might have in this case?

Reply:
Oh! This is actually defined behaviour.  

In this case, payload1 is accepted as an argument (key value).  

CommandParser's job is to screen for syntactically wrong inputs only. Setting payload as argument is a logical error on the users' part.   

As we discussed, the responsibility of logic-based checks like type-checking and length-checking is up to each feature using CommandParser to avoid tightly coupling the Parser to each feature.

This also follows the [Single Responsibility Principle](https://nus-cs2113-ay2223s2.github.io/website/se-book-adapted/common/definitions.html#s).   

A benefit of the hashmap architecture is that we can effectively ignore bad inputs like these.  
We can just `.get` the keys we expect to have.

If we want to be super strict, on the feature level, once an unknown key is encountered, the feature can just throw a custom `InvalidArgument` exception or handle the error along those lines. 

I don't see a way to differentiate arguments from payloads without each feature's parsing logic being tightly coupled with each feature's unique requirements. What are your thoughts?

Reply:
okay it makes sense. Maybe can just leave it to each feature to check for validity of arguments.